### ISSUE TITLE
.Include() with Self-Join / Coalesce

### ISSUE THREAD

**Issue by brent-russell** [aspnet/EntityFramework#3101]
Beta 7 and Beta 6.

This might encompass two separate issues. The Include() method is problematic with the following query:

```C#
var query = from eVersion in _dbContext.Entities
			join eRoot in _dbContext.Entities
				on eVersion.RootEntityId.GetValueOrDefault() equals eRoot.Id
			into RootEntities
			from eRootJoined in RootEntities.DefaultIfEmpty()  // left outer join
			select eRootJoined ?? eVersion;
```

If `query = query.Include(e => e.Children);` is used, the coalesce expression `eRootJoined ?? eVersion` cannot be bound.

If `Include` is added to the second or both of the `_dbContext.Entities` lines, then a null reference exception occurs.

If `Include` is added to the first `_dbContext.Entities` line, the query runs, but does not include the children.

```C#
public class Entity
{
	public Entity()
	{
		this.Children = new Collection<Child>();
	}

	public int Id { get; set; }

	public int? RootEntityId { get; set; }

	public Entity RootEntity { get; set; }

	public ICollection<Child> Children { get; set; }
}
```

Where RootEntityId is a self-referential relation on the Entity table.

**Comment by maumar** [comment]
@brent-russell  could you also share a model you are using?

**Comment by brent-russell** [comment]
@maumar edited original comment to include model info.

**Comment by maumar** [comment]
poaching this

**Comment by maumar** [comment]
blocked by #4547

**Comment by maumar** [comment]
blocked by #4925

**Comment by maumar** [comment]
clearing milestone for retriage

**Comment by rowanmiller** [comment]
@maumar where are we at with this issue?

**Comment by rowanmiller** [comment]
@maumar ping 😄

**Comment by maumar** [comment]
@rowanmiller fix is almost ready

**Comment by maumar** [comment]
Fixed in 7bdf5818b10389588d74ecc03ba766fe24850298
